### PR TITLE
test: Enable Spark 4.0 tests

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
## Which issue does this PR close?

Part of  https://github.com/apache/datafusion-comet/issues/372

## Rationale for this change

To be ready for Spark 4.0

## What changes are included in this PR?

This PR enables the spark-4.0 tests with comet enabled except for the ones listed in https://github.com/apache/datafusion-comet/issues/551

## How are these changes tested?

ANSI is enabled for Spark-4.0

### CODE REVIEW

**Comment by kazuyukitanimura** [comment]
@andygrove @comphead @huaxingao @parthchandra @viirya 
This is ready for review

File: .github/workflows/spark_sql_test_ansi.yml [hunk 1]
```diff
@@ -22,17 +22,17 @@ concurrency:
...
-  #  pull_request:
-  #    paths-ignore:
-  #      - "doc/**"
-  #      - "**.md"
-
-  # manual trigger ONLY
+  push:
+    paths-ignore:
+      - "doc/**"
+      - "docs/**"
```
Comment (comphead) [inline .github/workflows/spark_sql_test_ansi.yml:28]: we need to have only 1 docs folder..... its not related to this PR thoug

File: common/src/main/java/org/apache/comet/parquet/TypeUtil.java [hunk 2]
```diff
@@ -169,6 +170,7 @@ && isUnsignedIntTypeMatched(logicalTypeAnnotation, 64)) {
         break;
       case INT96:
         if (sparkType == TimestampNTZType$.MODULE$) {
+          if (isSpark40Plus()) return;
```
Comment (comphead) [inline common/src/main/java/org/apache/comet/parquet/TypeUtil.java:173]: wondering should we log something or do a blind return?

File: spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala [hunk 3]
```diff
@@ -32,6 +33,12 @@ trait CometExprShim {
         (unhex.child, Literal(false))
     }
 
+    protected def isTimestampNTZType(dt: DataType): Boolean = dt match {
```
Comment (comphead) [inline spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala:36]: can it be just 
```
dt.typeName == "timestamp_ntz" 
```
?

File: spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala [hunk 4]
```diff
@@ -32,5 +33,11 @@ trait CometExprShim {
         (unhex.child, Literal(false))
     }
 
+    protected def isTimestampNTZType(dt: DataType): Boolean = dt match {
```
Comment (comphead) [inline spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala:36]: ditto

File: .github/workflows/spark_sql_test_ansi.yml (hunk 1 above)
Comment (kazuyukitanimura) [reply .github/workflows/spark_sql_test_ansi.yml:27]: removed doc

File: spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala (hunk 3 above)
Comment (kazuyukitanimura) [reply spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala]: updated

File: spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala (hunk 4 above)
Comment (kazuyukitanimura) [reply spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala]: updated

File: spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala [hunk 5]
```diff
@@ -33,11 +33,8 @@ trait CometExprShim {
         (unhex.child, Literal(false))
     }
 
-    protected def isTimestampNTZType(dt: DataType): Boolean = dt match {
-        // `TimestampNTZType` is private
-        case dt if dt.typeName == "timestamp_ntz" => true
-        case _ => false
-    }
+    protected def isTimestampNTZType(dt: DataType): Boolean =
+        dt.typeName == "timestamp_ntz" => true // `TimestampNTZType` is private
```
Comment (comphead) [inline spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala:37]:
```suggestion
        dt.typeName == "timestamp_ntz" // `TimestampNTZType` is private
```

File: spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala [hunk 6]
```diff
@@ -33,11 +33,8 @@ trait CometExprShim {
         (unhex.child, Literal(false))
     }
 
-    protected def isTimestampNTZType(dt: DataType): Boolean = dt match {
-        // `TimestampNTZType` is private
-        case dt if dt.typeName == "timestamp_ntz" => true
-        case _ => false
-    }
+    protected def isTimestampNTZType(dt: DataType): Boolean =
+        dt.typeName == "timestamp_ntz" => true // `TimestampNTZType` is private
```
Comment (comphead) [inline spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala:37]:
```suggestion
        dt.typeName == "timestamp_ntz" // `TimestampNTZType` is private
```

File: spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala (hunk 5 above)
Comment (kazuyukitanimura) [reply spark/src/main/spark-3.2/org/apache/comet/shims/CometExprShim.scala]: Thanks, fixed

File: spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala (hunk 6 above)
Comment (kazuyukitanimura) [reply spark/src/main/spark-3.3/org/apache/comet/shims/CometExprShim.scala]: Thanks, fixed

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 7]
```diff
@@ -716,7 +716,7 @@ class CometSparkSessionExtensions
       // enabled.
       if (isANSIEnabled(conf)) {
         if (COMET_ANSI_MODE_ENABLED.get()) {
-          logWarning("Using Comet's experimental support for ANSI mode.")
+          logInfo("Using Comet's experimental support for ANSI mode.")
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:719]: Is this change necessary for Spark 4 support?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 8]
```diff
@@ -88,7 +87,7 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
       case _: BinaryType => 8
       case _: TimestampType => 9
       case _: DecimalType => 10
-      case dt if dt.typeName == "timestamp_ntz" => 11
+      case dt if isTimestampNTZType(dt) => 11
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:90]: Nice improvement

**Comment by andygrove** [comment]
I didn't review the Spark diff, but the other changes LGTM from a first pass. I will review more carefully tomorrow.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 7 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:719]: Without this change, the log gets really long and makes things hard to follow...

**Comment by kazuyukitanimura** [comment]
@andygrove just checking in to see if you have more feedback

**Comment by kazuyukitanimura** [comment]
Friendly ping @andygrove @comphead @huaxingao @parthchandra @viirya

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @kazuyukitanimura

**Comment by kazuyukitanimura** [comment]
Waiting for #581 gets merged first

**Review by comphead — APPROVED** [review]
lgtm, thanks @kazuyukitanimura 👍

File: .github/workflows/spark_sql_test_ansi.yml [hunk 9]
```diff
@@ -44,8 +42,8 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        java-version: [11]
-        spark-version: [{short: '3.4', full: '3.4.2'}]
+        java-version: [17]
+        spark-version: [{short: '4.0', full: '4.0.0-preview1'}]
```
Comment (viirya) [inline .github/workflows/spark_sql_test_ansi.yml:46]: Is this particular for ansi test? Hmm, is there any difference between 3.4 and 4.0?

Comment (kazuyukitanimura) [reply .github/workflows/spark_sql_test_ansi.yml:46]: This has been disabled. List of failing tests https://github.com/apache/datafusion-comet/issues/551

Comment (viirya) [reply .github/workflows/spark_sql_test_ansi.yml:46]: Hmm, before we can enable it, can we still use 3.4 to run these tests?

Comment (kazuyukitanimura) [reply .github/workflows/spark_sql_test_ansi.yml:46]: I have not tried running this with 3.4, but based on the comment, it likely fails.

Comment (viirya) [reply .github/workflows/spark_sql_test_ansi.yml:46]: Ah, I forgot we don't run ansi test in CI.

**Comment by kazuyukitanimura** [comment]
Merged thank you @andygrove @comphead @viirya
